R/vpc-simulation.R

```diff
@@ -0,0 +1,158 @@
+  # $PRIOR, $THETA/$THETAP/$THETAPV, $OMEGA/$OMEGAP/$OMEGAPD, $SIGMA/$SIGMAP/$SIGMAPD
+  records_rm <- c("prior", "theta", "thetapv", "omega", "omegapd", "sigma", "sigmapd")
+  purrr::walk(records_rm, function(rec){
+    if(mod_has_record(vpc_sim, type = rec)) remove_records(vpc_sim, type = rec)
+  })
+
+  # Set up simulation
+  setup_vpc_sim(vpc_sim, n = n, seed = seed)
+
+  # Return read-in model to get the updated class as part of the model object
+  return(read_model(file.path(model_dir, vpc_dir)))
+}
+
+
+get_msf_path <- function(.mod){
+  ctl <- safe_read_ctl(.mod)
+  ests <- nmrec::select_records(ctl, "est")
+
+  msf_path <- purrr::map(ests, function(est){
+    opt <- nmrec::get_record_option(est, "msf")
+    if(!is.null(opt)) opt$value else opt
+  }) %>% purrr::list_c()
+
+  if(!is.null(msf_path)){
+    if(length(msf_path) != 1){
+      msf_txt <- paste0(msf_path, collapse = ", ")
+      rlang::abort(
+        c(glue("Multiple MSF files found: {msf_txt}"))
+      )
+    }
+    msf_path <- file.path(.mod[[ABS_MOD_PATH]], msf_path)
+  }
+
+  return(msf_path)
+}
+
+
+#' Set up VPC simulation
+#'
+#' Creates new `$SIMULATION` and `$MSFI` records
+setup_vpc_sim <- function(.vpc_sim, n = 100, seed = 1234){
+  checkmate::assert_numeric(n, lower = 1)
+  checkmate::assert_numeric(seed)
+  # Check for $PRED or $ERROR records (cant be both)
+  after_opts <- c(
+    "pred" = mod_has_record(.vpc_sim, "pred"),
+    "error" = mod_has_record(.vpc_sim, "error")
+  )
+  if(all(after_opts)){
+    dev_error("setup_vpc_sim only supports $PRED or $ERROR records. Not both.")
+  }
+  after_name <- names(after_opts[after_opts])
+
+
+  ## Add new variable assignment to $ERROR _or_ $PRED: `REPI=IREP` ##
+  # Store current $error or $pred record
+  after_rec <- get_records(.vpc_sim, after_name)
+  after_rec <- after_rec[[length(after_rec)]]
+  after_lines <- after_rec$get_lines()
+  pred_error_lines <- after_lines[after_lines != ""]
+  # TODO: do we need to check if this was already defined?
+  pred_error_lines <- c(pred_error_lines, "REPI=IREP", "")
+
+  # Remove old record, and add a new one with the modification
+  #  - Use existing record name, dont overwrite
+  remove_records(.vpc_sim, after_name)
+  add_new_record(.vpc_sim, after_name, rec_name = NULL, lines = pred_error_lines, after = NULL)
+
+  ## Add new $SIMULATION record (must be after $ERROR or $PRED, but before $TABLE) ##
+  sim_lines <- glue("({seed}) ({seed} NORMAL) SUBPROBLEMS={n} TRUE=FINAL")
+  add_new_record(.vpc_sim, "simulation", lines = sim_lines, after = after_name)
+
+  ## Add new $TABLE record with predefined format and columns (including REPI) ##
+  table_name <- glue("{get_model_id(get_based_on(.vpc_sim))}sim.tab")
+  table_lines <- glue("ONEHEADER REPI DV PRED FILE={table_name}")
+  add_new_record(.vpc_sim, "table", lines = table_lines, after = "simulation")
+
+  ## Add new $MSFI record (end of file) ##
+  msf_path <- get_msf_path(read_model(get_based_on(.vpc_sim)))
```

Comment:
Wherever we're doing something like this, we should do `get_based_on(.vpc_sim)[1]` because that could potentially return a vector.